src/core/net/srp_server.cpp

```diff
@@ -0,0 +1,1500 @@
+
+    VerifyOrExit(aDnsHeader.GetAdditionalRecordsCount() == 2, error = OT_ERROR_FAILED);
+
+    // EDNS(0) Update Lease Option.
+
+    SuccessOrExit(error = Dns::Name::ReadName(aMessage, aOffset, aHeaderOffset, name, sizeof(name)));
+    SuccessOrExit(error = aMessage.Read(aOffset, leaseRecord));
+    VerifyOrExit(leaseRecord.IsValid(), error = OT_ERROR_PARSE);
+    aOffset += leaseRecord.GetSize();
+
+    aHost->SetLease(leaseRecord.GetLeaseInterval());
+    aHost->SetKeyLease(leaseRecord.GetKeyLeaseInterval());
+
+    // SIG(0).
+
+    sigOffset = aOffset;
+    SuccessOrExit(error = Dns::Name::ReadName(aMessage, aOffset, aHeaderOffset, name, sizeof(name)));
+    SuccessOrExit(error = aMessage.Read(aOffset, sigRecord));
+    VerifyOrExit(sigRecord.IsValid(), error = OT_ERROR_PARSE);
+
+    sigRdataOffset = aOffset + sizeof(Dns::ResourceRecord);
+    aOffset += sizeof(sigRecord);
+
+    // TODO: verify that the signature doesn't expire.
+
+    SuccessOrExit(error = Dns::Name::ReadName(aMessage, aOffset, aHeaderOffset, signerName, signerNameLength));
+
+    signatureLength = sigRecord.GetLength() - (aOffset - sigRdataOffset);
+    VerifyOrExit(signatureLength <= sizeof(signature), error = OT_ERROR_NO_BUFS);
+    VerifyOrExit(aMessage.ReadBytes(aOffset, signature, signatureLength) == signatureLength, error = OT_ERROR_PARSE);
+    aOffset += signatureLength;
+
+    // Verify the signature. Currently supports only ECDSA.
+
+    VerifyOrExit(sigRecord.GetAlgorithm() == Dns::KeyRecord::kAlgorithmEcdsaP256Sha256, error = OT_ERROR_FAILED);
+    VerifyOrExit(signatureLength == Crypto::Ecdsa::P256::Signature::kSize, error = OT_ERROR_PARSE);
+    VerifyOrExit(sigRecord.GetTypeCovered() == 0, error = OT_ERROR_FAILED);
+
+    SuccessOrExit(error = VerifySignature(*aHost->GetKey(), aMessage, aDnsHeader, sigOffset, sigRdataOffset,
+                                          sigRecord.GetLength()));
+
+exit:
+    return error;
+}
+
+otError Server::VerifySignature(const Dns::Ecdsa256KeyRecord &aKey,
+                                const Message &               aMessage,
+                                Dns::Header                   aDnsHeader,
+                                uint16_t                      aSigOffset,
+                                uint16_t                      aSigRdataOffset,
+                                uint16_t                      aSigRdataLength)
+{
+    otError                        error;
+    uint16_t                       offset = aMessage.GetOffset();
+    uint16_t                       signatureOffset;
+    Crypto::Sha256                 sha256;
+    Crypto::Sha256::Hash           hash;
+    Crypto::Ecdsa::P256::Signature signature;
+
+    VerifyOrExit(aSigRdataLength >= Crypto::Ecdsa::P256::Signature::kSize, error = OT_ERROR_INVALID_ARGS);
+
+    sha256.Start();
+
+    // SIG RDATA less signature.
+    sha256.Update(aMessage, aSigRdataOffset, aSigRdataLength - Crypto::Ecdsa::P256::Signature::kSize);
```

Comment:
I was struggling with it and tried to verify with uncompressed signer name but put it down to see if you have different interpreting. So it seems we should sign & verify with uncompressed form :)

I will make the change soon.

Reply:
To be honest, I would like/prefer if the wire format from message would be directly used for sig calculation (whether compressed or not). In my view, it would be simpler and easier to implement (in both server and client sides).

I think RFC-2335 requirs uncompressed name for sig calculation. I was hoping SIG(0) RFC-2931 would override that but from my reading/understanding of it, it seems not to.  I noticed that the implementation here [Abhayakara/mdnsresponder](https://github.com/Abhayakara/mdnsresponder/blob/main/ServiceRegistration/towire.c#L507) also seems to use uncompressed version for sig calculation but then append it as compressed.

Reply:
> To be honest, I would like/prefer if the wire format from message would be directly used for sig calculation (whether compressed or not). In my view, it would be simpler and easier to implement (in both server and client sides).
>
> I think RFC-2335 requirs uncompressed name for sig calculation. I was hoping SIG(0) RFC-2931 would override that but from my reading/understanding of it, it seems not to. 

I share the same idea and this is what I was struggling with. I guess the reason of using uncompressed form of the signer name is to include the actual signer name to produce the signature. For messages that are signed with selected RRs (covered-type), the actual signer name may not be included for creating the signature if uncompressed form is used. But this is not a problem for request signature, because it uses the request message starting from DNS header and the actual signer name will always be included even if uncompressed form is used. I think RFC-2931 is trying to be compatible with RFC 2535 on this portion.

Reply:
I think references to "RFC 2335" above should be "RFC 2535".

Yes, RFC 2535 specifies the canonical form to ensure that signature calculation is well-defined, especially in the presence of caching DNS servers. RFC 2931 leverages existing mechanisms and hence maintains usage of the canonical form.